### PR TITLE
[VOL-89] VolumeDriver.List for golang dvol plugin

### PR DESCRIPTION
**Pull request by robhaswell**
https://clusterhq.atlassian.net/browse/VOL-89

The simplest test and implementation for VolumeDriver.List in dvol.

In this branch I have moved the response handler out of the main function. Eventually all handlers should be expressed like this (possibly each in their own file).

I have also added some types into the dvol and datalayer APIs to make it clearer what is being communicated across the borders. There are also utility methods to get the volume path.

### CODE REVIEW

**Comment by lukemarsden** [comment]
I can imagine it would fail against docker version currently on travis, because 1.8.2 doesn't have a `docker volume` subcommand.

Please add a decorator which can check which version of Docker it's running against, and skip if it is a version of docker from before the docker volume API landed (`docker volume ls`, etc). I think this should be `if docker_version < 1.9, skip`.

File: dvol_python/testtools.py [hunk 1]
```diff
@@ -20,6 +22,17 @@
...
 
+def _skip_max_docker_ver(ver):
+    try:
+        return max_ver(DOCKER_VERSION, ver) == ver
+    except ValueError:
+        return False
+
+skip_if_docker_version_less_than = lambda ver: skipIf(
+    _skip_max_docker_ver(ver),
+    "Not expected to in this Docker version")
```
Comment (zubron) [inline dvol_python/testtools.py]: Missing word: "Not expected to *work* in this Docker version"

File: dvol_python/testtools.py [hunk 2]
```diff
@@ -20,6 +22,17 @@
...
 )
 
+def _skip_max_docker_ver(ver):
+    try:
+        return max_ver(DOCKER_VERSION, ver) == ver
+    except ValueError:
+        return False
+
+skip_if_docker_version_less_than = lambda ver: skipIf(
+    _skip_max_docker_ver(ver),
```
Comment (zubron) [inline dvol_python/testtools.py]: I think I'm reading this right... If `DOCKER_VERSION` is equal to `ver` then `_skip_max_docker_ver` will return `True` which means that the test will be skipped. But we only want to skip if `DOCKER_VERSION` is _less_ than `ver`.

File: dvol_python/testtools.py [hunk 3]
```diff
@@ -20,6 +22,17 @@
     "Not expected to work in Python version"
 )
 
+def _skip_max_docker_ver(ver):
+    try:
+        return max_ver(DOCKER_VERSION, ver) == ver
```
Comment (zubron) [inline dvol_python/testtools.py]: You could use `semver.compare` to check for differences between the versions.

**Comment by zubron** [comment]
Thanks @robhaswell! I don't think the comment about the checking of test versions will affect our tests since we test with version 1.10.2 but it would still be good to fix. Please address and then merge when green :smile:

**Comment by robhaswell** [comment]
Thanks for the review @zubron !

**Pull request merged**